Q.1

In [6]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

Q.2 Observations are data set records whereas variables are fields/columns in a data set.

Q.3

In [6]:
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [13]:
 print(df['survived'].value_counts())

survived
0    549
1    342
Name: count, dtype: int64


In [17]:
df.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

In [21]:
print(df['survived'].value_counts())
print(df['pclass'].value_counts())
print(df['sex'].value_counts())
print(df['age'].value_counts())
print(df['sibsp'].value_counts())
print(df['parch'].value_counts())
print(df['fare'].value_counts())
print(df['embarked'].value_counts())
print(df['class'].value_counts())
print(df['who'].value_counts())
print(df['adult_male'].value_counts())
print(df['deck'].value_counts())
print(df['embark_town'].value_counts())
print(df['alive'].value_counts())
print(df['alone'].value_counts())


survived
0    549
1    342
Name: count, dtype: int64
pclass
3    491
1    216
2    184
Name: count, dtype: int64
sex
male      577
female    314
Name: count, dtype: int64
age
24.00    30
22.00    27
18.00    26
19.00    25
28.00    25
         ..
36.50     1
55.50     1
0.92      1
23.50     1
74.00     1
Name: count, Length: 88, dtype: int64
sibsp
0    608
1    209
2     28
4     18
3     16
8      7
5      5
Name: count, dtype: int64
parch
0    678
1    118
2     80
5      5
3      5
4      4
6      1
Name: count, dtype: int64
fare
8.0500     43
13.0000    42
7.8958     38
7.7500     34
26.0000    31
           ..
35.0000     1
28.5000     1
6.2375      1
14.0000     1
10.5167     1
Name: count, Length: 248, dtype: int64
embarked
S    644
C    168
Q     77
Name: count, dtype: int64
class
Third     491
First     216
Second    184
Name: count, dtype: int64
who
man      537
woman    271
child     83
Name: count, dtype: int64
adult_male
True     537
False    354
Name: count, dtype: int64

Q.4 The discrepancy between df.shape and df.describe() is due to the fact that df.shape reflects the overall number of variables and observations wheras df.describe() includes only numerical data.

Q.5 The difference between attributes and methods is that attributes store information, whereas methods perform specific actions.

In [25]:
df.shape

(891, 15)

In [24]:
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

Q.6 The df.describe() method provides the 'count', 'mean', 'std', 'min', '25%', '50%', '75%', and 'max' summary statistics for each variable it analyzes. Give the definitions (perhaps using help from the ChatBot if needed) of each of these summary statistics
  Each of these values are summary statistics used to find the distibution of values in a data set.
Count shows the amount of existing values in each numerical column. The mean is the average value of the values in a column, calculated by dividing the sum of all the existing values in the column by the count of the values. STD is the standard deviation of the data points, meaning the dispersion of the data points in reference to the mean.
The max and min are the highest and lowest value in each column representing numerical data. 25%, 50% and 75% are quartiles into which the data points can be organized. The first quartile (25%) is the point before which a quarter of the data points lie, the median (50%) is the middle point of the data when all the values are organized from smallest to largest, finally, the third quartile (75%), is the point before which 75% of the data points lie.

Q.7
1. Provide an example of a "use case" in which using df.dropna() might be peferred over using del df['col']
  A use case where df.dropna() would be preferred over del df[‘col’] is where a potential client of a marketing company was missing mandatory information such as contact information, so their information is unusable, so the entire client record would be deleted.  

2. Provide an example of "the opposite use case" in which using del df['col'] might be preferred over using df.dropna() 
  A situation in which del df[‘col’] would be preferred is if a newspaper company discontinued one of its papers so information about whether each client is subscribed to it or not is no longer valid and would be deleted.

3. Discuss why applying del df['col'] before df.dropna() when both are used together could be important
  To avoid deleting records that have missing values in the column that is meant to be deleted. 

4. Remove all missing data from one of the datasets you're considering using some combination of del df['col'] and/or df.dropna() and give a justification for your approach, including a "before and after" report of the results of your approach for your dataset.
  To remove the missing data from my dataset, I need to identify which column has a lot of missing values. Using df.describe() I can see that the age column has 714 values as opposed to the rest with 891. When I delete all the missing values using df.dropna(), I am left with 182 rows, so I should first delete the age column and then the rest of the rows with missing values. When I first delete age and then all rows with missing values, I am left with 201 rows and 14 columns.

In [26]:
dfna=df.dropna()

In [27]:
dfna.shape

(182, 15)

In [29]:
dfna.head(5)

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
1          1       1  female  38.0      1      0  71.2833        C  First   
3          1       1  female  35.0      1      0  53.1000        S  First   
6          0       1    male  54.0      0      0  51.8625        S  First   
10         1       3  female   4.0      1      1  16.7000        S  Third   
11         1       1  female  58.0      0      0  26.5500        S  First   

      who  adult_male deck  embark_town alive  alone  
1   woman       False    C    Cherbourg   yes  False  
3   woman       False    C  Southampton   yes  False  
6     man        True    E  Southampton    no   True  
10  child       False    G  Southampton   yes  False  
11  woman       False    C  Southampton   yes   True

In [39]:
del df['age']

In [40]:
df.shape

(891, 14)

In [11]:
df.head(5)

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [41]:
df.dropna()

survived  pclass     sex  sibsp  parch     fare embarked  class    who  \
1           1       1  female      1      0  71.2833        C  First  woman   
3           1       1  female      1      0  53.1000        S  First  woman   
6           0       1    male      0      0  51.8625        S  First    man   
10          1       3  female      1      1  16.7000        S  Third  child   
11          1       1  female      0      0  26.5500        S  First  woman   
..        ...     ...     ...    ...    ...      ...      ...    ...    ...   
871         1       1  female      1      1  52.5542        S  First  woman   
872         0       1    male      0      0   5.0000        S  First    man   
879         1       1  female      0      1  83.1583        C  First  woman   
887         1       1  female      0      0  30.0000        S  First  woman   
889         1       1    male      0      0  30.0000        C  First    man   

     adult_male deck  embark_town alive  alone  
1         False    C    Cherbourg   yes  False  
3         False    C  Southampton   yes  False  
6          True    E  Southampton    no   True  
10        False    G  Southampton   yes  False  
11        False    C  Southampton   yes   True  
..          ...  ...          ...   ...    ...  
871       False    D  Southampton   yes  False  
872        True    B  Southampton    no   True  
879       False    C    Cherbourg   yes  False  
887       False    B  Southampton   yes   True  
889        True    C    Cherbourg   yes   True  

[201 rows x 14 columns]

Q.8 

In [18]:
df.groupby("age")["sex"].describe().head(20)

count unique     top freq
age                            
0.42      1      1    male    1
0.67      1      1    male    1
0.75      2      1  female    2
0.83      2      1    male    2
0.92      1      1    male    1
1.00      7      2    male    5
2.00     10      2  female    6
3.00      6      2    male    4
4.00     10      2  female    5
5.00      4      1  female    4
6.00      3      2  female    2
7.00      3      2    male    2
8.00      4      2  female    2
9.00      8      2  female    4
10.00     2      2  female    1
11.00     4      2    male    3
12.00     1      1    male    1
13.00     2      1  female    2
14.00     6      2  female    4
14.50     1      1  female    1

Q.8.1
This command takes the values under colum 1, in this case age, and groups them by the frequency of each value. Then the values from collum 2, in this case sex, are organized into the existing groups from column 1. This allows for a representation of the frequency of values under specific variables in relation to each other, the missing values are exluded to provide an accurate representation.

Q.8.2
The discrepencies between the count in df.describe() and df.groupby are due to the fact that df.groupby excludes rows with missing values under the chosen columns for accurate grouping. The command df.describe gives a general understanding of how many rows have existing values under each individual column, this allows us to see whether missing values are present. 

Q.8.3
Upon running the incorrect code and using both google and chatbot to troubleshoot, I found that the chatbot is a much more direct search engine for the solution, as google will sometimes directly provide a solution but most of the time multiple sources would have to be used.   

HW 1
pre lecture chatbot summary
    Here is a summary of our exchanges, as requested, to help with your homework assignment:
You started by asking for a way to calculate and display value counts for all columns in your dataset [value counts for all columns]. The initial response provided Python code using a for loop to iterate through each column and apply the .value_counts() method, along with explanations based on the sources. You then asked for a way to achieve the same result using a single statement [value counts for all columns in one statement], to which the response offered a solution using the .apply() method with pd.value_counts().
Most recently, you inquired about the difference between "attributes" and "methods" in pandas DataFrames, using df.shape and df.describe() as examples [What is the difference between an "attribute", such as df.shape which does not end with () and a "method", such as df.describe() which does end with ()]. This reflects a progression from requesting specific code snippets to seeking a deeper understanding of fundamental programming concepts relevant to data analysis with pandas, as covered in the sources.
Your questions and the provided responses highlight several key points from the source material:
●
Value Counts and Data Types: The sources emphasize that while .value_counts() can be applied to both numerical and categorical data, it is generally more meaningful for categorical columns. This distinction is important for choosing appropriate summary statistics based on the nature of your data.
●
Missing Values: The default behavior of .value_counts() (and several other pandas methods) regarding missing values is a recurring theme in the sources and our conversation. You've explored how to include missing values in counts using dropna=False and how missingness can affect other summary statistics.
●
Methods vs. Attributes: Your question about methods and attributes touches on the distinction between actions that modify or operate on a DataFrame (methods) and properties that describe the DataFrame (attributes). This difference is important for understanding how to interact with and manipulate DataFrames effectively.
The progression of your queries suggests a developing understanding of core data analysis concepts in Python, moving from basic syntax to more nuanced distinctions between data types, handling missingness, and the roles of attributes and methods in pandas.

Post lecture chatbot summary
    Here is a summary of our conversation since you last requested a summary for your homework assignment:
●
You asked for a description of the output generated by the pandas function describe.
●
I provided an explanation of the output of df.describe(), including the meaning of each statistic: count, mean, std, min, 25% (Q1), 50% (Median), 75% (Q3), and max. I emphasized that .describe() only works on numerical columns, and that it handles missing values by excluding them from calculations.
●
You then asked for the difference between df.dropna() and del df['col'].
●
I explained that df.dropna() is used to remove rows containing missing values, while del df['col'] is used to completely delete a specified column. I noted that the choice between the two depends on whether you want to remove rows with missing data or irrelevant columns. I also discussed the efficiency of using del df['col'] before df.dropna() to avoid unnecessarily removing rows.
●
Finally, you asked for an explanation of the summary statistics count, mean, std, min, 25%, 50%, 75%, and max.
●
I provided a detailed explanation of each statistic, drawing on the sources you provided and our conversation history. I emphasized the importance of understanding how missing values are handled, and how these statistics help characterize the distribution of numerical data.

Chatbot summary for Q.8


written parts:
Q.2 Observations are data set records whereas variables are fields/columns in a data set.

Q.4 The discrepancy between df.shape and df.describe() is due to the fact that df.shape reflects the overall number of variables and observations wheras df.describe() includes only numerical data.

Q.5 The difference between attributes and methods is that attributes store information, whereas methods perform specific actions.

Q.7

Provide an example of a "use case" in which using df.dropna() might be peferred over using del df['col']
A use case where df.dropna() would be preferred over del df[‘col’] is where a potential client of a marketing company was missing mandatory information such as contact information, so their information is unusable, so the entire client record would be deleted. 

Provide an example of "the opposite use case" in which using del df['col'] might be preferred over using df.dropna() 
A situation in which del df[‘col’] would be preferred is if a newspaper company discontinued one of its papers so information about whether each client is subscribed to it or not is no longer valid and would be deleted.

Discuss why applying del df['col'] before df.dropna() when both are used together could be important
To avoid deleting records that have missing values in the column that is meant to be deleted. 

Remove all missing data from one of the datasets you're considering using some combination of del df['col'] and/or df.dropna() and give a justification for your approach, including a "before and after" report of the results of your approach for your dataset.
To remove the missing data from my dataset, I need to identify which column has a lot of missing values. Using df.describe() I can see that the age column has 714 values as opposed to the rest with 891. When I delete all the missing values using df.dropna(), I am left with 182 rows, so I should first delete the age column and then the rest of the rows with missing values. When I first delete age and then all rows with missing values, I am left with 201 rows and 14 columns.